src/librustdoc/visit_ast.rs

```diff
@@ -61,20 +61,63 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
             krate.item.span,
             &Spanned { span: rustc_span::DUMMY_SP, node: hir::VisibilityKind::Public },
             hir::CRATE_HIR_ID,
             &krate.item.module,
             None,
         );
-        // Attach the crate's exported macros to the top-level module:
-        module.macros.extend(krate.exported_macros.iter().map(|def| (def, None)));
-        module.is_crate = true;
+        top_level_module.is_crate = true;
+        // Attach the crate's exported macros to the top-level module.
+        // In the case of macros 2.0 (`pub macro`), and for built-in `derive`s or attributes as
+        // well (_e.g._, `Copy`), these are wrongly bundled in there too, so we need to fix that by
+        // moving them back to their correct locations.
+        'exported_macros: for def in krate.exported_macros {
+            // The `def` of a macro in `exported_macros` should correspond to either:
+            //  - a `#[macro-export] macro_rules!` macro,
+            //  - a built-in `derive` (or attribute) macro such as the ones in `::core`,
+            //  - a `pub macro`.
+            // Only the last two need to be fixed, thus:
+            if def.ast.macro_rules {
+                top_level_module.macros.push((def, None));
+                continue 'exported_macros;
+            }
+            let tcx = self.cx.tcx;
+            /* Because of #77828 we cannot do the simpler:
+            let macro_parent_module = tcx.def_path(tcx.parent_module(def.hir_id).to_def_id());
+            // and instead have to do: */
+            let macro_parent_module = tcx.def_path({
+                use rustc_middle::ty::DefIdTree;
+                tcx.parent(tcx.hir().local_def_id(def.hir_id).to_def_id()).unwrap()
+            });
+            // HACK: rustdoc has no way to lookup `doctree::Module`s by their HirId. Instead,
+            // lookup the module by its name, by looking at each path segment one at a time.
+            let mut cur_mod = &mut top_level_module;
+            for path_segment in macro_parent_module.data {
+                let path_segment_ty_ns = match path_segment.data {
+                    rustc_hir::definitions::DefPathData::TypeNs(symbol) => symbol,
+                    _ => {
+                        // If the path segment is not from the type namespace
+                        // (_e.g._, it can be from a value namespace in the case of `f::` in:
+                        // `fn f() { pub macro m() {} }`
+                        // then the item is not accessible, and should thus act as if it didn't
+                        // exist (unless "associated macros" (inside an `impl`) were a thing…).
+                        continue 'exported_macros;
+                    }
+                };
+                cur_mod = cur_mod
+                    .mods
+                    .iter_mut()
+                    .find(|module| module.name == Some(path_segment_ty_ns))
+                    .unwrap();
```

Comment:
I can imagine this breaking on something like

```rust
pub mod a {
  pub macro m() {}
}
pub use a as b;
pub use b::m;
```
if `def_path()` returned `b::m` instead of `a::m`. I don't think it will ever do that, but could you add the test case?

Reply:
Except for some differences in `pub`, we already have:

https://github.com/rust-lang/rust/blob/a59bcb1c910437f35c57ccdd55025d044b207440/src/test/rustdoc/macro_pub_in_module.rs#L46-L52

Could `pub` affect HIR `def_path`s?

Reply:
Okay, I think I've managed to make this hack fail:

```rust
enum Crazy {
    Weird = {
        pub macro but_why() {}
        42
    },
}
```

This manages to hit a parent in the type namespace which is not `mod`ule.

Reply:
Oh nice, good catch 😆 I don't think that needs to be supported here, but it would be nice to have a test to show the current behavior.